src/System.Net.Sockets/src/System/Net/Sockets/SafeSocketHandle.cs

```diff
@@ -171,6 +177,11 @@ internal void CloseAsIs(bool abortive)
                 Dispose();
                 if (innerSocket != null)
                 {
+                    // In case we cancel operations, switch to an abortive close.
+                    // Unless the user requested a normal close using Socket.Shutdown.
+                    bool canceledOperations = false;
+                    bool canAbort = abortive || !_sentShutdown;
```

Comment:
I'm having trouble following what abortive vs canAbort means.  Can you clarify?

Reply:
Before this PR, `TryUnblockSocket` always performed an abortive (that is :_Connection reset by peer_) close.
We're changing to take into account if the user has called `Socket.Shutdown`.
`canAbort` tells `TryUnblockSocket` if it can do an abortive close.

Reply:
> Before this PR, TryUnblockSocket always performed an abortive close.

But TryUnblockSocket was already taking a `bool abortive` argument.  If it always performed an abortive close, what was the purpose of that argument?

Reply:
There is a special case for Sockets that  don't have the CLOEXEC flag set:

https://github.com/dotnet/corefx/blob/485441f79ead7878af206a6af61a300e1458616a/src/System.Net.Sockets/src/System/Net/Sockets/SafeSocketHandle.Unix.cs#L379-L381

Reply:
I'm still struggling here, @tmds.  Are you saying that both `canAbort` and `abortive` mean "do an abortive close", but they communicate that information to different parts of the stack?  The names are too close to be meaningful.

Reply:
I missed this comment.

`canAbort` contains whether a `CloseAsIs` that was called with `abortive: false`, can become a an `innerSocket.Close` with `abortive: true`.

Reply:
I'll remove `canAbort` and replace it with `!hasShutdownSend`.